0011-rendezvous-hashing.md

```diff
@@ -0,0 +1,311 @@
+
+# Improvements
+
+In this ADR we attempt to improve the resharding experience by minimizing the
+difference between different sharding topologies.
+
+## Better Overlap Between Sharding Topologies
+
+Currently, two different sharding topologies, even if they share nodes, will
+have a small overlap between addressing schemas. This can be significantly
+improved by using a different sharding algorithm.
+
+For this purpose we replace the implementation of
+`ShardingBlobAccessConfiguration` with one that uses [Rendezvous
+hashing](https://en.wikipedia.org/wiki/Rendezvous_hashing). Rendezvous hashing
+is a lightweight and stateless technique for distributed hash tables. It has a
+low overhead with minimal disruption during resharding.
+
+Sharding with Rendezvous hashing gives us the following properties:
+ * Removing a shard is _guaranteed_ to only require resharding for the blobs
+   that resolved to the removed shard.
+ * Adding a shard will reshard any blob to the new shard with a probability of
+   `weight/total_weight`.
+
+This effectively means adding or removing a shard triggers a predictable,
+minimal amount of resharding, eliminating the need for drained backends.
+
+```
+message ShardingBlobAccessConfiguration {
+  message Shard {
+    // unchanged
+    BlobAccessConfiguration backend = 1;
+    // unchanged
+    uint32 weight = 2;
+  }
+  // NEW:
+  // Shards identified by a key within the context of this sharding
+  // configuration. The key is a freeform string which describes the identity
+  // of the shard in the context of the current sharding configuration.
+  // Shards are chosen via Rendezvous hashing based on the digest, weight and
+  // key of the configuration.
+  //
+  // When removing a shard from the map it is guaranteed that only blobs
+  // which resolved to the removed shard will get a different shard. When
+  // adding shards there is a weight/total_weight probability that any given
+  // blob will be resolved to the new shards.
+  map<string, Shard> shards = 2;
+  // NEW:
+  message Legacy {
+    // Order of the shards for the legacy schema. Each key here refers to
+    // a corresponding key in the 'shard_map' or null for drained backends.
+    repeated string shard_order = 1;
+    // Hash initialization seed used for legacy schema.
+    uint64 hash_initialization = 2;
+  }
+  // NEW:
+  // A temporary legacy mode which allows clients to use storage backends which
+  // are sharded with the old sharding topology implementation. Consumers are
+  // expected to migrate in a timely fashion and support for the legacy schema
+  // will be removed by 2025-12-31.
+  Legacy legacy = 3;
+}
+```
+
+## Migration Instructions
+
+If you are fine with dropping your data and repopulating your cache you can
+start using the new sharding algorithm right away. For clarity we have an
+example on how to perform a non-destructive migration.
+
+Given the following configuration showing an unmirrored sharding topology of two
```

Comment:
Probably good to add "###" headings above each of these configurations, so that this section becomes easier to follow?